op-supervisor/supervisor/backend/db/db_test.go

```diff
@@ -49,11 +52,23 @@ func TestChainsDB_Rewind(t *testing.T) {
 
 type stubHeadStorage struct{}
 
+func (s *stubHeadStorage) Apply(heads.Operation) error {
+	panic("not implemented")
+}
+
+func (s *stubHeadStorage) Current() *heads.Heads {
+	panic("not implemented")
+}
+
 type stubLogDB struct {
 	addLogCalls  int
 	headBlockNum uint64
 }
 
+func (s *stubLogDB) LastCheckpointBehind(entrydb.EntryIdx) (*logs.Iterator, error) {
+	panic("not implemented")
+}
```

Comment:
iterators getting into the tests is actually a big challenge -- if the thing you're testing is using an iterator, the requirements for testing cascade until you've got a real database...

So either we interface and stub out the Iterator (which could be ok) *or* we rethink how the iterator is exposed to other components. Will try the former since it allows mocking `NextLog`, which is helpful. But this could be indicating a smell where the internals of the database are being inappropriately exposed.

@protolambda thoughts?

Reply:
yeah this is getting me to rethink the way I've put together the update function. Thinking now it should be more like:

* ChainsDB is responsible for asking each logDB to attempt to advance, providing the existing local head as a limit
* LogDB can use the existing starting point to construct an iterator, and then can return `NextExecutingMessage`
* ChainsDB can use the checker to check each ExecutingMessage that it passes

This hides the iterator from the ChainsDB, simplifying the testing and making responsibilities more clear.

Reply:
For locking and iterator-persistence purposes though, I guess even the checking should happen in the LogDB. So the checker can be passed down to the LogDB

Reply:
I solved this by adding `NextExecutingMessage` to the LogDB. This advances the iterator and returns the EM at the same time, and is easy to stub out for testing.